docs/ai-coder/tasks-core-principles.md

```diff
@@ -0,0 +1,164 @@
+- **Productivity:** Eliminate manual processes from developer processes enabling them to focus on less defined and harder-to-do issues
+- **Scalability:** Once a workflow is captured in a task, it can be reused by other teams within your organization scaling with you as you grow
+- **Flexibility:** Support both developer _AND_ autonomous agentic workflows
+
+### Example Task Workflow
+
+![Example Background Task](../images/guides/ai-agents/background-task-example.png)Example of Background Coder Tasks operation.
+
+## How to Make a Task Template
+
+### Refresher: What are Templates
+
+As a quick refresher, a template defines the underlying infrastructure that a Coder workspace runs on. Templates themselves are written in Terraform managed as a `main.tf` file that defines the contents of the workspace and the resources it requires to run. Templates can also pull in Dockerfiles, other build files, and startup scripts or config files for special configuration.
+
+Within this configuration, Coder specifically looks for
+
+- `coder_agent`: Coder's fundamental resource that runs inside a workspace to enable connectivity to external systems
+- `coder_workspace`: Defines the workspace owner, state, etc.
+- `coder_provisioner`: Defines basic system information like system architecture, OS, etc.
+- `coder_app`: Pull in IDE or other applications like VS Code, terminals, web apps
+
+Coder templates also supports standard Terraform providers for connecting to external systems like AWS/Azure/GCP.
+
+### What Makes a Task Template
+
+Task templates are regular Coder Templates, with a few specific resources defined additionally. These resources prime the template and corresponding workspaces for automated execution and AI-driven workflows rather than development environments for developers and builders.
+
+The specific resources that turn a template into a task template include:
+
+- **AgentAPI Module:** Coder's task execution engine that provides Web UI integration, task reporting, and agent lifecycle management that makes any module compatible with Coder Tasks
+- `coder_parameter` named _ai_prompt_: Define the AI prompt input so users can define/specify what tasks need to run
+
+The following code snippet can be dropped into any existing template to modify it into a Claude-Code enabled task template. This snippet also includes space for a setup script that will prime the agent for execution.
+
+```hcl
+data "coder_parameter" "ai_prompt" {
+    name = "AI Prompt"
+    type = "string"
+}
+
+data "coder_parameter" "setup_script" {
+  name         = "setup_script"
+  display_name = "Setup Script"
+  type         = "string"
+  form_type    = "textarea"
+  description  = "Script to run before running the agent"
+  mutable      = false
+  default      = ""
+}
+
+# The Claude Code module does the automatic task reporting
+# Other agent modules: https://registry.coder.com/modules?search=agent
+# Or use a custom agent:  
+module "claude-code" {
```

Comment:
Use the latest example from the registry. We are on version 3.0 now, and there is a change in variable names.